File: app/src/main/java/eu/szwiec/replayview/replay/ReplayFragment.java
package eu.szwiec.replayview.replay;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;
import com.github.angads25.filepicker.model.DialogConfigs;
import com.github.angads25.filepicker.model.DialogProperties;
import com.github.angads25.filepicker.view.FilePickerDialog;

import java.io.File;

import eu.szwiec.replayview.R;
import eu.szwiec.replayview.databinding.FragmentReplayBinding;
import rm.com.youtubeplayicon.PlayIconDrawable;

import static rm.com.youtubeplayicon.PlayIconDrawable.IconState.PAUSE;
import static rm.com.youtubeplayicon.PlayIconDrawable.IconState.PLAY;

public class ReplayFragment extends Fragment {

    private ReplayViewModel mViewModel;
    private FragmentReplayBinding mBinding;

    private PlayIconDrawable mPlayPauseIconDrawable;
    private MaterialDialog mProgressDialog;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_replay, container, false);
        mBinding.setLifecycleOwner(this);

        View view = mBinding.getRoot();

        mViewModel = ViewModelProviders.of(this).get(ReplayViewModel.class);

        mViewModel.getIsProcessingLiveData().observe(this, isProcessing -> {
            if (isProcessing) {
                if(mProgressDialog == null) {
                    mProgressDialog = buildProgressDialog();
                }
                mProgressDialog.show();
            } else {
                mProgressDialog.dismiss();
            }
        });

        mViewModel.getEventsLiveData().observe(this, events -> {
            if (events.size() != 0) {
                enablePlayackControls(true);
            } else {
                enablePlayackControls(false);
                Toast.makeText(getContext(), "No data", Toast.LENGTH_SHORT).show();
            }

            mBinding.totalTime.setText(mViewModel.getTotalTime());
        });

        mViewModel.getIsPlayingLiveData().observe(this, isPlaying -> {
            if (isPlaying) {
                mPlayPauseIconDrawable.animateToState(PAUSE);
            } else {
                mPlayPauseIconDrawable.animateToState(PLAY);
            }
        });

        mViewModel.getProgressLiveData().observe(this, progress -> {
            mBinding.seekbar.setProgress(progress);
        });

        init();

        return view;
    }

    private void init() {
        mPlayPauseIconDrawable = buildPlayPause();
        enablePlayackControls(true);

        mBinding.playPauseButton.setOnClickListener(v -> mViewModel.toggle());
        mBinding.pickFileButton.setOnClickListener(v -> buildDataTypeDialog().show());

        mBinding.seekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if(fromUser) {
                    mViewModel.setProgress(progress);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    private MaterialDialog buildProgressDialog() {
        return new MaterialDialog.Builder(this.getContext())
                .content("Loading...")
                .progress(true, 0)
                .cancelable(false)
                .build();
    }

    private PlayIconDrawable buildPlayPause() {
        return PlayIconDrawable.builder()
                .withInterpolator(new FastOutSlowInInterpolator())
                .withDuration(300)
                .withInitialState(PLAY)
                .into(mBinding.playPauseButton);
    }


    private FilePickerDialog buildFilePickerDialog() {
        DialogProperties properties = new DialogProperties();

        properties.selection_mode = DialogConfigs.SINGLE_MODE;
        properties.selection_type = DialogConfigs.FILE_SELECT;
        properties.root = Environment.getExternalStorageDirectory();
        properties.error_dir = new File(DialogConfigs.DEFAULT_DIR);
        properties.offset = new File(DialogConfigs.DEFAULT_DIR);
        properties.extensions = new String[]{"zip"};

        FilePickerDialog filePickerDialog = new FilePickerDialog(getContext(), properties);
        filePickerDialog.setTitle("Select a File");

        filePickerDialog.setDialogSelectionListener(files -> {
            mViewModel.zipPicked(files[0]);
        });

        return filePickerDialog;
    }

    private MaterialDialog buildDataTypeDialog() {

        return new MaterialDialog.Builder(getContext())
                .title("Choose data type")
                .items(mViewModel.availableDataTypes)
                .itemsCallbackMultiChoice(null, (dialog, which, text) -> {
                    mViewModel.typesPicked(text);
                    return true;
                })
                .positiveText("Choose")
                .onPositive((dialog, which) -> {
                    buildFilePickerDialog().show();
                })
                .build();
    }

    private void enablePlayackControls(boolean enabled) {
        mBinding.playPauseButton.setEnabled(enabled);
        mBinding.seekbar.setEnabled(enabled);
        mBinding.speedButton.setEnabled(enabled);

        if (enabled) {
            mPlayPauseIconDrawable.setColor(getResources().getColor(R.color.colorPrimaryDark));
        } else {
            mPlayPauseIconDrawable.setColor(getResources().getColor(R.color.dbc_light_grey));
        }
    }
}

File: app/src/main/java/eu/szwiec/replayview/replay/ReplayFragment.kt
package eu.szwiec.replayview.replay

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.os.Environment
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.afollestad.materialdialogs.MaterialDialog
import com.github.angads25.filepicker.model.DialogConfigs
import com.github.angads25.filepicker.model.DialogProperties
import com.github.angads25.filepicker.view.FilePickerDialog
import eu.szwiec.replayview.App
import eu.szwiec.replayview.R
import eu.szwiec.replayview.databinding.FragmentReplayBinding
import org.jetbrains.anko.AnkoLogger
import java.io.File
import javax.inject.Inject

class ReplayFragment : Fragment(), AnkoLogger {

    @Inject
    lateinit var viewModelFactory: ReplayViewModelFactory

    private lateinit var viewModel: ReplayViewModel
    private lateinit var binding: FragmentReplayBinding

    private val progressDialog: MaterialDialog by lazy { buildProgressDialog() }
    private val typePickerDialog: MaterialDialog by lazy { buildTypePickerDialog() }
    private val filePickerDialog: FilePickerDialog by lazy { buildFilePickerDialog() }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        (context?.applicationContext as App).component.inject(this)

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_replay, container, false)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(ReplayViewModel::class.java)
        binding.let {
            it.viewModel = viewModel
            it.setLifecycleOwner(this)
        }

        val view = binding.root

        viewModel.stateLD.observe(this, Observer { state ->
            when (state) {
                State.PICKING_TYPE -> typePickerDialog.show()
                State.PICKING_FILE -> {
                    filePickerDialog.show()
                    filePickerDialog.findViewById<View>(R.id.select).isEnabled = true //for testing, but in the perfect world, I should get rid of it
                }
                State.PROCESSING -> progressDialog.show()
                State.READY -> progressDialog.dismiss()
                State.ERROR -> {
                    progressDialog.dismiss()
                    Toast.makeText(activity, "Something went wrong", Toast.LENGTH_SHORT).show()
                }
            }
        })

        return view
    }

    private fun buildProgressDialog(): MaterialDialog {
        return MaterialDialog.Builder(this.context!!)
                .content("Loading...")
                .progress(true, 0)
                .cancelable(false)
                .build()
    }

    private fun buildFilePickerDialog(): FilePickerDialog {
        val properties = DialogProperties()

        properties.selection_mode = DialogConfigs.SINGLE_MODE
        properties.selection_type = DialogConfigs.FILE_SELECT
        properties.root = Environment.getExternalStorageDirectory()
        properties.error_dir = File(DialogConfigs.DEFAULT_DIR)
        properties.offset = File(DialogConfigs.DEFAULT_DIR)
        properties.extensions = arrayOf("zip")

        val filePickerDialog = FilePickerDialog(context, properties)
        filePickerDialog.setTitle(getString(R.string.file_picker_title))

        filePickerDialog.setDialogSelectionListener { files -> viewModel.onFilePicked(files.firstOrNull()) }
        filePickerDialog.setOnDismissListener({ viewModel.dialogDismissed() })

        return filePickerDialog
    }

    private fun buildTypePickerDialog(): MaterialDialog {
        return MaterialDialog.Builder(context!!)
                .title(getString(R.string.type_picker_title))
                .items(AVAILABLE_TYPES)
                .itemsCallbackMultiChoice(null) { dialog, which, types ->
                    viewModel.setPickedTypes(types)
                    true
                }
                .dismissListener { viewModel.dialogDismissed() }
                .positiveText("Choose")
                .onPositive { dialog, which -> viewModel.onTypePicked() }
                .build()
    }
}